### PR TITLE
[nbs] getMany read planning and parallelize

### PR DESCRIPTION
**Pull request by max-hoffman**
Add read planning for journal getMany so instead of doing randIO we get seqIO. SeqIO has a large effect for disk storage systems. Also parallelize getMany calls. The biggest change here is the semantics around how reading locks the journal file. Reads hold the journal lock for longer, which could lower write throughput in some cases. If we see evidence of this, we can do more work to limit the amount of time batch reads can interruptibly holding the journal lock.

### CODE REVIEW

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_median | to_latency_median | is_faster |
|---------------------|---------------------|-------------------|-----------|
| tpcc-scale-factor-1 | 223.34              | 130.13            | 1         |


|      test_name      | server_name |              server_version              |  tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|-------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | a6ce23920dbde470bac74b9ffd0da5fe9848eec9 | 14.19 | tpcc-scale-factor-1 | dolt        | c3af9848b2ddf389d03d6dc5ca381b36640ab560 | 16.71 | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>c3af984</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>c3af984</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 3.02                | 2.97              | 0         |
| groupby_scan         | 17.63               | 17.63             | 0         |
| index_join           | 5.28                | 5.18              | 0         |
| index_join_scan      | 2.3                 | 2.3               | 0         |
| index_scan           | 55.82               | 55.82             | 0         |
| oltp_point_select    | 0.54                | 0.54              | 0         |
| oltp_read_only       | 8.58                | 8.58              | 0         |
| select_random_points | 0.84                | 0.84              | 0         |
| select_random_ranges | 0.99                | 0.99              | 0         |
| table_scan           | 55.82               | 55.82             | 0         |
| types_table_scan     | 164.45              | 161.51            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.79                | 6.79              | 0         |
| oltp_insert           | 3.36                | 3.43              | 0         |
| oltp_read_write       | 16.41               | 16.41             | 0         |
| oltp_update_index     | 3.49                | 3.49              | 0         |
| oltp_update_non_index | 3.43                | 3.43              | 0         |
| oltp_write_only       | 7.84                | 7.98              | 0         |
| types_delete_insert   | 7.56                | 7.56              | 0         |

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f47c827</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f47c827</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_median | to_latency_median | is_faster |
|---------------------|---------------------|-------------------|-----------|
| tpcc-scale-factor-1 | 84.47               | 90.78             | 0         |


|      test_name      | server_name |              server_version              | tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | 5d5847fe97134da54a453fe27cf20627dee59abc | 0.47 | tpcc-scale-factor-1 | dolt        | f47c8271ca49f8e7086835125836c231092aa06f | 17.03 | -1        |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 3.02                | 3.02              | 0         |
| groupby_scan         | 17.63               | 17.95             | 0         |
| index_join           | 5.28                | 5.28              | 0         |
| index_join_scan      | 2.26                | 2.3               | 0         |
| index_scan           | 54.83               | 54.83             | 0         |
| oltp_point_select    | 0.53                | 0.53              | 0         |
| oltp_read_only       | 8.58                | 8.58              | 0         |
| select_random_points | 0.84                | 0.84              | 0         |
| select_random_ranges | 0.99                | 0.99              | 0         |
| table_scan           | 55.82               | 55.82             | 0         |
| types_table_scan     | 161.51              | 158.63            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 6.79                | 6.79              | 0         |
| oltp_insert           | 3.36                | 3.43              | 0         |
| oltp_read_write       | 16.41               | 16.41             | 0         |
| oltp_update_index     | 3.49                | 3.49              | 0         |
| oltp_update_non_index | 3.43                | 3.43              | 0         |
| oltp_write_only       | 7.84                | 7.84              | 0         |
| types_delete_insert   | 7.56                | 7.56              | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>ba06fd4</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>ba06fd4</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8d0834c</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8d0834c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>6699fca</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>6699fca</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/store/nbs/journal_chunk_source.go [hunk 1]
```diff
@@ -69,44 +72,78 @@ func (s journalChunkSource) get(_ context.Context, h hash.Hash, _ *Stats) ([]byt
...
+	// idx is the array offset into the shared |reqs|
+	idx int
 }
 
-func (s journalChunkSource) getManyCompressed(ctx context.Context, _ *errgroup.Group, reqs []getRecord, found func(context.Context, CompressedChunk), stats *Stats) (bool, error) {
+func (s journalChunkSource) getMany(ctx context.Context, eg *errgroup.Group, reqs []getRecord, found func(context.Context, *chunks.Chunk), stats *Stats) (bool, error) {
+	return s.getManyCompressed(ctx, eg, reqs, func(ctx context.Context, cc CompressedChunk) {
+		ch, err := cc.ToChunk()
+		if err != nil {
+			eg.Go(func() error {
```
Comment (reltuk) [inline go/store/nbs/journal_chunk_source.go:86]: Unfortunately, I don't think this is race-free. In particular, I don't think it's safe to `eg.Go()` on an errgroup where `eg.Wait()` may have already been called.

**Review by reltuk — COMMENTED** [review]
This looks good to me, other than the race in getMany. I don't know how to measure the impact on the lock duration here vs. what we used to have before...it's something we can always come back and fix though. I think this is good to go if we fix the race.

Simplest way to fix the race might be to move getManyCompressed impl to a helper function that has a signature something like:

```go
func (s journalChunkSource) getManyCompressed_ext(ctx context.Context, eg *errgroup.Group, reqs []getRecord, found func(context.Context, CompressedChunk) error, stats *Stats) (bool, error) {
```

and then to wrap it in both `getMany` and `getManyCompressed`...

File: go/store/nbs/journal_chunk_source.go (hunk 1 above)
Comment (reltuk) [reply go/store/nbs/journal_chunk_source.go:86]: Actually, I think I might be wrong here...in particular, errgroup is implemented in terms of a sync.WaitGroup, and I thought all positive delta wg.Add() calls were dangerous after a wg.Wait(), but that's not true...

```
// Note that calls with a positive delta that occur when the counter is zero
// must happen before a Wait. Calls with a negative delta, or calls with a
// positive delta that start when the counter is greater than zero, may happen
// at any time.
```

Comment (reltuk) [reply go/store/nbs/journal_chunk_source.go:86]: And by, "I think I might be wrong here"...I'm quite sure I'm wrong and your usage is safe / to spec. Sorry for the noise.

**Comment by max-hoffman** [comment]
We've talked about how this may negatively impact mixed workloads that will now block on reads for longer. Tradeoff being that batch reads that will complete faster, but are less interruptible because the read holds the journal lock for longer. The tradeoff for tablescans is pretty obvious, if we see issues with write throughput after adding this we add changes that make batch reads only hold the journal lock for short periods of time.

**Pull request merged**